File: android/automated-tests/src/androidTest/kotlin/br/com/zup/beagle/automatedTests/cucumber/steps/TabBarSteps.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.automatedTests.cucumber.steps

import androidx.test.rule.ActivityTestRule
import br.com.zup.beagle.automatedTests.R
import br.com.zup.beagle.automatedTests.activity.MainActivity
import br.com.zup.beagle.automatedTests.cucumber.robots.ScreenRobot
import br.com.zup.beagle.automatedTests.utils.ActivityFinisher
import br.com.zup.beagle.automatedTests.utils.TestUtils
import cucumber.api.java.After
import cucumber.api.java.Before
import cucumber.api.java.en.Given
import cucumber.api.java.en.Then
import cucumber.api.java.en.When
import org.junit.Rule

const val TAB_BAR_SCREEN_BFF_URL = "http://10.0.2.2:8080/tabbar"
const val TAB_BAR_SCREEN_HEADER = "TabBar Screen"

class TabBarSteps {

    @Rule
    var activityTestRule = ActivityTestRule(MainActivity::class.java)

    @Before("@tabBar")
    fun setup() {
        TestUtils.startActivity(activityTestRule, TAB_BAR_SCREEN_BFF_URL)
    }

    @After("@tabBar")
    fun tearDown() {
        ActivityFinisher.finishOpenActivities()
    }

    @Given("^that I'm on the tabBar screen$")
    fun checkTabBarScreen() {
        ScreenRobot()
            .checkViewContainsText(TAB_BAR_SCREEN_HEADER, true)
    }

    @When("I click in a tab with text (.*)")
    fun clickOnTab(text:String){
        ScreenRobot()
            .clickOnText(text)
    }

    @Then("^a tabBarItem with text (.*) should exist$")
    fun checksTabItemsExists(text: String){
        ScreenRobot()
            .checkViewContainsText(text)
    }

    @Then("^the tab position should have its text changed to (.*)$")
    fun checksSetContextResult(text: String){
        ScreenRobot()
            .checkViewContainsText(text)
    }

    @Then("^I click on each tabBarItem and confirm its position$")
    fun clickOnText() {
        ScreenRobot()
            .clickOnText("Tab1").checkViewContainsText("Tab position 0")
            .clickOnText("Tab2").checkViewContainsText("Tab position 1")
            .clickOnText("Tab3").checkViewContainsText("Tab position 2")
            .clickOnText("Tab4").checkViewContainsText("Tab position 3")
            .clickOnText("Tab5").checkViewContainsText("Tab position 4")
            .clickOnText("Tab6").checkViewContainsText("Tab position 5")
            .clickOnText("Tab7").checkViewContainsText("Tab position 6")
            .clickOnText("Tab8").checkViewContainsText("Tab position 7")
            .clickOnText("Tab9").checkViewContainsText("Tab position 8")
            .clickOnText("Tab10").checkViewContainsText("Tab position 9")
    }

    @Then("^the tab with text (.*) must be on screen$")
    fun checkTabBarRendersTabs(text:String) {
        ScreenRobot()
            .checkViewContainsText(text)
    }

    @Then("^check tab with text (.*) and beagle icon are on screen$")
    fun checkTitleAndIconOnTab(title:String){
        ScreenRobot()
            .checkTabContainsTextAndIcon(title, R.drawable.beagle)
    }

    @Then("^check tab with beagle icon is on screen$")
    fun checkTabWithIconOnly(){
        ScreenRobot()
            .checkTabContainsTextAndIcon(icon = R.drawable.beagle)
    }

    @Then("^the tab with text (.*) and beagle icon will change icon to delete icon$")
    fun checkTabWithIconChanged(title:String){
        ScreenRobot()
            .checkTabContainsTextAndIcon(title = title, icon = android.R.drawable.ic_delete)
    }
}

File: iOS/Tests/AutomatedTestsUITests/Steps/TabBarSteps.swift
/*
* Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import Foundation
import XCTest

// swiftlint:disable implicitly_unwrapped_optional
// swiftlint:disable force_unwrapping
class TabBarSteps: CucumberStepsDefinition {
    
    var application: XCUIApplication!
    
    func loadSteps() {
            
        before { scenarioDefinition in
            if scenarioDefinition?.tags.contains("tabbar") ?? false {
                let url = "http://localhost:8080/tabbar"
                self.application = TestUtils.launchBeagleApplication(url: url)
            }
        }
        
        Given("^that I'm on the tabBar screen$") { _, _ -> Void in
            XCTAssertTrue(TabBarScreenElements.SCREEN_HEADER.element.exists)
        }

        // MARK: - When

        // Scenario 2, 6
        When(#"^I click on button "([^\"]*)" to change tab$"#) { args, _ -> Void in
            let text = args![0]
            self.application.buttons[text].tap()
        }

        // Scenario 3
        When(#"^I click in a tab with text "([^\"]*)"$"#) { args, _ -> Void in
            let text = args![0]
            self.application.staticTexts[text].tap()
        }
        
        // MARK: - Then
        
        // Scenario 1
        Then(#"^I click on each tabBarItem and confirm its position$"#) { _, _ -> Void in
            for i in 0...9 {
                let tabTitle = "Tab\(i + 1)"
                let positionText = "Tab position \(i)"
                self.application.staticTexts[tabTitle].tap()
                XCTAssertTrue(self.application.staticTexts[positionText].exists)
            }
        }

        // Scenario 2
        Then(#"^the tab with text "([^\"]*)" must be on screen$"#) { args, _ -> Void in
            let text = args![0]
            XCTAssertTrue(self.application.staticTexts[text].exists)
        }
        
        // Scenario 3
        Then(#"^the tab position should have its text changed to "([^\"]*)"$"#) { args, _ -> Void in
            let text = args![0]
            XCTAssertTrue(self.application.staticTexts[text].exists)
        }
        
        // Scenario 4, 6
        Then(#"^check tab with "([^\"]*)" icon is on screen$"#) { args, _ -> Void in
            let icon = args![0]
            XCTAssertTrue(self.application.images[icon].exists)
        }

        // Scenario 5
        Then(#"^check tab with text "([^\"]*)" and "([^\"]*)" icon are on screen$"#) { args, _ -> Void in
            let text = args![0]
            let icon = args![1]
            XCTAssertTrue(self.application.staticTexts[text].exists)
            XCTAssertTrue(self.application.images[icon].exists)
        }
        
        // Scenario 6
        Then(#"^the tab with text "([^\"]*)" and "([^\"]*)" icon will change icon to "([^\"]*)" icon$"#) { args, _ -> Void in
            let text = args![0]
            let changedIcon = args![2]
            XCTAssertTrue(self.application.staticTexts[text].exists)
            XCTAssertTrue(self.application.images[changedIcon].exists)
        }

    }

    private enum TabBarScreenElements: String {
        case SCREEN_HEADER = "TabBar Screen"
        
        var element: XCUIElement {
            return XCUIApplication().staticTexts[self.rawValue]
        }
    }
}
